### ISSUE TITLE
Contextual type inference breaks on function parameters with explicit return

### ISSUE THREAD

**Issue by evankim390** [microsoft/TypeScript#53954]
# Bug Report

### 🔎 Search Terms

contextual type inference function return generic

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about generics.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.0.4#code/JYOwLgpgTgZghgYwgAgPIAczAPYgM4A8A0gHzIDeAUMjcgNYQCeAXMkQNzW0wgAUDLNgEpWcEI04BfSpRgBXEAiy5kwACbESvbJhz5WvIcgC8ZDMvyajVWsjAALKNgDuyAOQLgAW3QAbCF4Q4BBqblIyCLh4YMgARlAQcHR4JqpqvIYmZDa0CWByUCAUXLY0AqzkaawAzABMyJIANCWlPPxM1i2luRD5hfRMAHTqg2DYAMpgUKAA5oac3TRNLZJSQpyUkfgxzthQyanqGUamyLw5ZUwVVch1Dc22bQKdi8h5BUUCw2qjE1Oz8xWzUkQnWQA)

### 💻 Code

```ts
interface Options<K> {
    key: K;
    fn(key: K): any;
}

function id<K>(options: () => Options<K>) {
    throw 'unimplemented';
}

const breaks = id(() => {
    return {
        key: { id: 32 },
        fn(key) {
            // 'key' is of type 'unknown'.(18046)
            return key.id.toString();
        },
    };
});

const works = id(() => ({
    key: { id: 32 },
    fn(key) {
        return key.id.toString();
    },
}));
```

### 🙁 Actual behavior

The `fn` parameter type is only correctly inferred in the case where an arrow function is directly returning an object.

### 🙂 Expected behavior

I would expect for both cases to be handled consistently.

**Comment by RyanCavanaugh** [comment]
Duplicate #47599

**Comment by jcalz** [comment]
@RyanCavanaugh Is a block body context sensitive in a way that a concise body isn’t?  I suspect as much but I’d love to hear official word since #47599 doesn’t seem to talk about that specifically.

**Comment by ahejlsberg** [comment]
This isn't related to #47599. Rather, it's an issue that was already fixed in #50903. The example works as expected in 5.1.

**Comment by ahejlsberg** [comment]
So, I'll call this a duplicate of #50687.
